chapter3/src/3_6_1/main.rs

```diff
@@ -0,0 +1,45 @@
+use std::io::{BufRead, BufReader, Result};
+
+const SOURCE: &str = r#"1行目
+2行目
+3行目
+"#;
+
+/// 終端文字を気にしながら1行ずつ読み込むパターン。
+fn with_eof() -> Result<()> {
+    let mut reader = BufReader::new(SOURCE.as_bytes());
+
+    loop {
+        let mut buf = String::new();
+        // `read_line` は `EOF` を検知してもエラーにすることはなく、
+        // `Ok(0)` が返ってきた時点で `EOF` を検知したことになる。
+        let num_bytes = reader.read_line(&mut buf)?;
+        print!("{}", buf);
+
+        if num_bytes == 0 {
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+/// 終端を気にせずもっと短く書きたい場合。
+fn without_eof() -> Result<()> {
+    let reader = BufReader::new(SOURCE.as_bytes());
+    let lines = reader.lines();
+
+    for line in lines.into_iter() {
```

Comment:
[nits] `for x in xs` は、 `xs` に対して `into_iter()` を呼び出すシンタックスシュガーなので、 ここは

```suggestion
    for line in lines {
```

と書けます。